Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -5289,6 +5295,178 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addFusedIfCompare(OpType op, Expression
+#endif
+        break;
+    default:
+        RELEASE_ASSERT_NOT_REACHED_WITH_MESSAGE("Op type '%s' is not a unary comparison.\n", makeString(opType).characters());
+    }
+
+    if (val1.isConst())
+        emitMoveConst(val1, resultLocation);
+    else
+        emitMove(val1.type(), val1Location, resultLocation);
+
+    skipVal1.link(&m_jit);
+
+    LOG_DEDENT();
+    return { };
+}
+
+PartialResult BBQJIT::addFusedSelectCompare(OpType opType, ExpressionType left, ExpressionType right, ExpressionType val1, ExpressionType val2, ExpressionType& result)
+{
+    if (left.isConst() && right.isConst()) {
+        BranchFoldResult foldResult = tryFoldFusedBranchCompare(opType, left, right);
+        Value src = (foldResult == BranchAlwaysTaken) ? val1 : val2;
+        Location srcLocation;
+        if (src.isConst())
+            result = src;
+        else {
+            result = topValue(val1.type());
+            srcLocation = loadIfNecessary(src);
+        }
+
+        LOG_INSTRUCTION("SelectCompare", makeString(opType).characters(), left, right, val1, val2, RESULT(result));
+        consume(left);
+        consume(right);
+        consume(val1);
+        consume(val2);
+        if (!result.isConst()) {
+            Location resultLocation = allocate(result);
+            emitMove(val1.type(), srcLocation, resultLocation);
+        }
+        return { };
+    }
+
+    Location leftLocation, rightLocation;
+    Location val1Location = Location::none(), val2Location = Location::none();
+
+    if (!left.isConst())
+        leftLocation = loadIfNecessary(left);
+    else if (left.isFloat())
+        emitMove(left, leftLocation = Location::fromFPR(wasmScratchFPR));
+
+    if (!right.isConst())
+        rightLocation = loadIfNecessary(right);
+    else if (right.isFloat()) {
+        if (!leftLocation.isFPR() || leftLocation.asFPR() != wasmScratchFPR)
+            emitMove(right, rightLocation = Location::fromFPR(wasmScratchFPR));
+    }
+
+    if (!val1.isConst())
+        val1Location = loadIfNecessary(val1);
+    if (!val2.isConst())
+        val2Location = loadIfNecessary(val2);
+
+    ASSERT(val1.isConst() || val1Location.isRegister());
+    ASSERT(val2.isConst() || val2Location.isRegister());
+
+    consume(val1);
+    consume(val2);
+
+    result = topValue(val1.type());
+    Location resultLocation = allocate(result);
+
+    LOG_INSTRUCTION("SelectCompare", makeString(opType).characters(), left, leftLocation, right, rightLocation, val1, val1Location, val2, val2Location, RESULT(result));
+    LOG_INDENT();
+
+    bool inverted = false;
+    if (val2Location == resultLocation) {
+        std::swap(val1, val2);
+        std::swap(val1Location, val2Location);
+        inverted = true;
+    }
+
+    // If the comparison operands alias with the result, we need to preserve them
+    // since we'll overwrite the result with val1 before doing the comparison
```

Comment:
You can just allocate the result before `consume`ing `val1/2` and avoid all this logic.